cram/cram.go

```diff
@@ -0,0 +1,488 @@
+	0x0f, 0xe0, 0x45, 0x4f, 0x46, 0x00, 0x00, 0x00, // |..EOF...|
+	0x00, 0x01, 0x00, 0x05, 0xbd, 0xd9, 0x4f, 0x00, // |......O.|
+	0x01, 0x00, 0x06, 0x06, 0x01, 0x00, 0x01, 0x00, // |........|
+	0x01, 0x00, 0xee, 0x63, 0x01, 0x4b, /*       */ // |...c.K|
+}
+
+// ErrNoEnd is returned when a stream cannot seek to a CRAM EOF block.
+var ErrNoEnd = errors.New("cram: cannot determine offset from end")
+
+// HasEOF checks for the presence of a CRAM magic EOF block.
+// The magic block is defined in the CRAM specification. A magic block
+// is written by a Writer on calling Close. The ReaderAt must provide
+// some method for determining valid ReadAt offsets.
+func HasEOF(r io.ReaderAt) (bool, error) {
+	type sizer interface {
+		Size() int64
+	}
+	type stater interface {
+		Stat() (os.FileInfo, error)
+	}
+	type lenSeeker interface {
+		io.Seeker
+		Len() int
+	}
+	var size int64
+	switch r := r.(type) {
+	case sizer:
+		size = r.Size()
+	case stater:
+		fi, err := r.Stat()
+		if err != nil {
+			return false, err
+		}
+		size = fi.Size()
+	case lenSeeker:
+		var err error
+		size, err = r.Seek(0, 1)
+		if err != nil {
+			return false, err
+		}
+		size += int64(r.Len())
+	default:
+		return false, ErrNoEnd
+	}
+
+	b := make([]byte, len(cramEOFmarker))
+	_, err := r.ReadAt(b, size-int64(len(cramEOFmarker)))
+	if err != nil {
+		return false, err
+	}
+	return bytes.Equal(b, cramEOFmarker), nil
+}
+
+// Reader is a CRAM format reader.
+type Reader struct {
+	r io.Reader
+
+	d definition
+	c *Container
+
+	err error
+}
+
+// NewReader returns a new Reader.
+func NewReader(r io.Reader) (*Reader, error) {
+	cr := Reader{r: r}
+	err := cr.d.readFrom(r)
+	if err != nil {
+		return nil, err
+	}
+	return &cr, nil
+}
+
+// Next advances the Reader to the next CRAM container. It returns false
+// when the stream ends, either by reaching the end of the stream or
+// encountering an error.
+func (r *Reader) Next() bool {
+	if r.err != nil {
+		return false
+	}
+	if r.c != nil {
+		io.Copy(ioutil.Discard, r.c.blockData)
+	}
+	var c Container
+	r.err = c.readFrom(r.r)
+	r.c = &c
+	return r.err == nil
+}
+
+// Container returns the current CRAM container. The returned Container
+// is only valid after a previous call to Next has returned true.
+func (r *Reader) Container() *Container {
+	return r.c
+}
+
+// Err returns the most recent error.
+func (r *Reader) Err() error {
+	if r.err == io.EOF {
+		return nil
+	}
+	return r.err
+}
+
+// definition is a CRAM file definition.
+//
+// See CRAM spec section 6.
+type definition struct {
+	Magic   [4]byte `is:"CRAM"`
+	Version [2]byte
+	ID      [20]byte
+}
+
+// readFrom populates a definition from the given io.Reader. If the magic
+// number of the file is not "CRAM" readFrom returns an error.
+func (d *definition) readFrom(r io.Reader) error {
+	err := binary.Read(r, binary.LittleEndian, d)
+	if err != nil {
+		return err
+	}
+	magic := reflect.TypeOf(*d).Field(0).Tag.Get("is")
```

Comment:
this seems like a complicated way to get `[]byte("CRAM")`

Reply:
At the time, it helped me keep the ownership details under control. If it's a performance problem we can revisit.